Jint/Native/Promise/PromiseInstance.cs

```diff
@@ -0,0 +1,124 @@
+using System.Threading.Tasks;
+using Jint.Native.Function;
+using Jint.Native.Object;
+using Jint.Runtime;
+using Jint.Runtime.Descriptors;
+using Jint.Runtime.Interop;
+
+namespace Jint.Native.Promise
+{
+    public class PromiseInstance : ObjectInstance
+    {
+        private readonly FunctionInstance _promiseResolver;
+        private readonly TaskCompletionSource<JsValue> _tcs = new TaskCompletionSource<JsValue>();
+
+        public Task<JsValue> Task => _tcs.Task;
+        public PromiseState State { get; private set; }
+
+        internal PromiseInstance(Engine engine) : base(engine, objectClass: "Promise")
+        {
+
+        }
+        
+        public PromiseInstance(Engine engine, FunctionInstance promiseResolver)
+            : this(engine)
+        {
+            _promiseResolver = promiseResolver;
+        }
+
+        public PromiseInstance(Engine engine, Task wrappedTask)
+            : this(engine)
+        {
+            wrappedTask.ContinueWith(t =>
+            {
+                if (t.Status == TaskStatus.RanToCompletion)
+                {
+                    var returnValue = Undefined;
+
+                    //  If the task returns a value
+                    var taskType = t.GetType();
+                    var resultProperty = taskType.GetProperty("Result", BindingFlags.Instance | BindingFlags.Public);
+
+                    if (resultProperty != null && resultProperty.PropertyType.Name != "VoidTaskResult")
+                        returnValue = FromObject(_engine, resultProperty.GetValue(t));
+                    
+                    _tcs.SetResult(returnValue);
+                    return;
+                }
+
+                _tcs.SetException(new PromiseRejectedException(FromObject(Engine, t.Exception?.InnerExceptions.FirstOrDefault() ?? new Exception("An unhandled exception was thrown"))));
+            });
+        }
+
+        public static PromiseInstance CreateResolved(Engine engine, JsValue result)
+        {
+            var resolved = new PromiseInstance(engine)
+            {
+                Prototype = engine.Promise.PrototypeObject,
+                Extensible = true
+            };
+
+            resolved._tcs.SetResult(result);
+            resolved.State = PromiseState.Resolved;
+
+            return resolved;
+        }
+
+        public static PromiseInstance CreateRejected(Engine engine, JsValue result)
+        {
+            var rejected = new PromiseInstance(engine)
+            {
+                Prototype = engine.Promise.PrototypeObject,
+                Extensible = true
+            };
+
+            rejected._tcs.SetException(new PromiseRejectedException(result));
+            rejected.State = PromiseState.Rejected;
+
+            return rejected;
+        }
+
+        internal void InvokePromiseResolver()
+        {
+            _promiseResolver.Invoke(new ClrFunctionInstance(_engine, "", Resolve, 1), new ClrFunctionInstance(_engine, "", Reject, 1));
```

Comment:
`ClrFunctionInstance`s might be reusable, like a field?

Reply:
This method can only be called once per promise instance so no advantage in caching these as fields.